### PR TITLE
DOP-4414: Lambda function to handle automated test deploys

### PR DESCRIPTION
**Pull request by anabellabuckvar**
### Stories/Links:

[DOP-4414](https://jira.mongodb.org/browse/DOP-4414)

### Notes

This PR includes a new webhook that is triggered by workflow action updates. However, there is logic in docs-worker-pool such that the smoke test deploy lambda will only be triggered, and therefore, only run, upon successful completion of a Deploy Staging ECS push job to main.

This PR also includes some refactoring and simplification of preexisting code for slack deploys and feature branch push jobs.

Please take note of error handling and the status codes returned for different cases and leave feedback if applicable! 

### Testing
Webhook, endpoint, and lambda were all tested by pushing to this branch and therefore triggering the workflow. Github pushes on feature branches were also tested  and verified to still be working as expected; although this PR did not make meaningful changes to that functionality, it was incorporated in refactoring

Cloud Docs Build example:
[Build Logs]( https://workerpoolstaging-qgeyp.mongodbstitch.com/pages/job.html?collName=dotcom_queue&jobId=65fc57846fc5f04850aacdd5)
[Build URL](https://mongodbcom-cdn.website.staging.corp.mongodb.com/docs-qa/atlas)

### README updates

- - [ ] This PR introduces changes that should be reflected in the README, and I have made those updates.
- - [x] This PR does not introduce changes that should be reflected in the README

### CODE REVIEW

File: api/controllers/v2/github.ts [hunk 1]
```diff
@@ -87,6 +141,138 @@ export const TriggerBuild = async (event: APIGatewayEvent): Promise<APIGatewayPr
...
+        body.workflow_run.head_branch +
+        ' will not trigger site deployments as it was not on master branch in upstream repo',
+    };
+  }
+
+  //automated test builds will always deploy in dotcomstg
+  const env = 'dotcomstg';
+
+  async function createAndInsertJob() {
+    for (const s in SMOKETEST_SITES) {
```
Comment (branberry) [inline api/controllers/v2/github.ts:194]: NIT: You can use the `of` keyword instead of `in` to get the value instead of the index i.e. `for (const repoName in SMOKEST_SITES)` would get you the value right away instead of having to index into the array

Comment (branberry) [inline api/controllers/v2/github.ts:194]: You could refactor the `for` loop and use `SMOKTEST_SITES.map` to allow for concurrent requests. Right now, these will occur sequentially since the for loop will `await` at each iteration at various points. If we use the map method built into arrays, we can get back an array of `Promise<void>` which we then can call `await Promise.all()`. This will make things a bit quicker 👍

File: src/repositories/projectsRepository.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * Project information from docs_metadata.projects for parser builds.
+ */
+export class ProjectsRepository extends BaseRepository {
+  constructor(db: Db, config: IConfig, logger: ILogger) {
+    super(config, logger, 'ProjectsRepository', db.collection('projects'));
+  }
+
+  async getProjectEntry(name: string): Promise<any> {
```
Comment (branberry) [inline src/repositories/projectsRepository.ts:14]: If we can, it would be great to properly type the return type. I know that the types are tied to the `BaseRepository`, so it might be a good chunk of work to make that possible. If that is the case, then no worries about having to do the work of updating all the types.

File: api/controllers/v2/github.ts [hunk 3]
```diff
@@ -1,71 +1,125 @@
...
+  };
+}
+
+export const triggerSmokeTestAutomatedBuild = async (event: APIGatewayEvent): Promise<APIGatewayProxyResult | null> => {
   const client = new mongodb.MongoClient(c.get('dbUrl'));
   await client.connect();
   const db = client.db(c.get('dbName'));
   const consoleLogger = new ConsoleLogger();
   const jobRepository = new JobRepository(db, c, consoleLogger);
+  const projectsRepository = new ProjectsRepository(client.db('docs_metadata'), c, consoleLogger);
```
Comment (branberry) [inline api/controllers/v2/github.ts:122]: I think it would be good to put the DB name `'docs_metadata'` as a const variable in case that we use this repository somewhere else. This would make sure we don't accidentally misspell the name of the database somewhere else.

File: api/controllers/v2/github.ts [hunk 4]
```diff
@@ -87,6 +141,138 @@ export const TriggerBuild = async (event: APIGatewayEvent): Promise<APIGatewayPr
...
+      statusCode: 202,
+      headers: { 'Content-Type': 'text/plain' },
+      body:
+        'Workflow ' +
+        body.workflow_run.name +
+        ' completed successfully but only Deploy Staging ECS workflow completion will trigger smoke test site deployments',
+    };
+
+  // if the build was not building master, no need for smoke test sites
+  if (body.workflow_run.head_branch != 'main' || body.repository.fork) {
```
Comment (branberry) [inline api/controllers/v2/github.ts:178]: Good idea to check for forks, I did not think of that!

**Review by branberry — COMMENTED** [review]
Hi Anabella, thank you so much for your work on this! Looks great. I don't have any big concerns, just lots of small stuff. Great work.

**Comment by branberry** [comment]
Hey Anabella, just one more thing. I do think it would be worth updating the README so that we know of this new process that occurs.

File: src/repositories/projectsRepository.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import { IConfig } from 'config';
+import { BaseRepository } from './baseRepository';
+import { ILogger } from '../services/logger';
+
+/**
+ * Project information from docs_metadata.projects for parser builds.
+ */
+export class ProjectsRepository extends BaseRepository {
+  constructor(db: Db, config: IConfig, logger: ILogger) {
+    super(config, logger, 'ProjectsRepository', db.collection('projects'));
```
Comment (mmeigs) [inline src/repositories/projectsRepository.ts:11]: Like @branberry mentioned above with `docs_metadata`, could we make `projects` a constant like in the other ____Repository classes

**Review by mmeigs — COMMENTED** [review]
Just one comment, and co-sign on all Brandon said! :) good work!

File: api/controllers/v2/github.ts (hunk 1 above)
Comment (anabellabuckvar) [reply api/controllers/v2/github.ts:194]: Will do, thanks for the idea!

File: api/controllers/v2/github.ts [hunk 6]
```diff
@@ -87,6 +141,138 @@ export const TriggerBuild = async (event: APIGatewayEvent): Promise<APIGatewayPr
...
+      const repoName = SMOKETEST_SITES[s];
+      const jobTitle = 'Smoke Test ' + repoName;
+      let repoInfo, projectEntry, repoOwner;
+
+      try {
+        repoInfo = await docsetsRepository.getRepo(repoName);
+        projectEntry = await projectsRepository.getProjectEntry(repoInfo.project);
+        repoOwner = projectEntry.github.organization;
+      } catch {
+        return 'repoInfo, projectEntry, or repoOwner not found for docs site ' + repoName;
```
Comment (mmeigs) [inline api/controllers/v2/github.ts:204]: nit: Could we implement some really basic logic here to try to narrow down the possibilities of the error? 
It could either use quick conditionals to find what is undefined, or we could simply try to log them all out within a template literal that shows which one is the issue.

File: api/controllers/v2/github.ts [hunk 7]
```diff
@@ -87,6 +141,138 @@ export const TriggerBuild = async (event: APIGatewayEvent): Promise<APIGatewayPr
...
+      }
+
+      const jobPrefix = repoInfo?.prefix ? repoInfo['prefix'][env] : '';
+
+      //add commit hash here: const newHead = body.workflow_run.head_sha;
```
Comment (mmeigs) [inline api/controllers/v2/github.ts:209]: Just wondering what this is? A future ticket or something that can be erased?

File: src/repositories/projectsRepository.ts (hunk 2 above)
Comment (anabellabuckvar) [reply src/repositories/projectsRepository.ts:14]: I added in a proper return type, but it's worth noting that nearly all of the methods in the `repositories` directory have a return type of `any`

File: api/controllers/v2/github.ts (hunk 7 above)
Comment (anabellabuckvar) [reply api/controllers/v2/github.ts:207]: Future ticket! [DOP-4401](https://github.com/mongodb/docs-worker-pool/pull/1003). I can remove the comment now but it didn't seem like an issue to leave in since I'm already working on that future ticket

File: src/repositories/projectsRepository.ts (hunk 5 above)
Comment (anabellabuckvar) [reply src/repositories/projectsRepository.ts:11]: @branberry @mmeigs to confirm, you mean like with the example of the docsets collection where it's set in `worker-env-construct.ts`from aws parameter store as shown [here](https://github.com/mongodb/docs-worker-pool/blob/025a8722b5aa69da322b4fbbc6cdee05f3d510e7/cdk-infra/lib/constructs/worker/worker-env-construct.ts#L67)?

File: api/controllers/v2/github.ts (hunk 6 above)
Comment (anabellabuckvar) [reply api/controllers/v2/github.ts]: I changed it so that they are all logged in a template literal for simplicity, lmk if it looks good to you!

File: api/controllers/v2/github.ts (hunk 1 above)
Comment (anabellabuckvar) [reply api/controllers/v2/github.ts]: I implemented this, please let me know if it's what you had in mind!

File: api/controllers/v2/github.ts [hunk 8]
```diff
@@ -1,71 +1,125 @@
...
+    jobType = 'productionDeploy';
+    branchName = 'master';
+  } else {
+    if (!githubEvent) {
+      throw 'Non SmokeTest Deploy jobs must have a github Event';
+    }
+    action = 'push';
+    jobType = 'githubPush';
+    branchName = githubEvent.ref.split('/')[2];
+    url = githubEvent?.repository.clone_url;
```
[the remark this answers is not in the document]
Comment (anabellabuckvar) [reply api/controllers/v2/github.ts]: Done

File: api/controllers/v2/github.ts [hunk 9]
```diff
@@ -1,71 +1,125 @@
...
-      isFork: githubEvent.repository.fork,
-      repoOwner: githubEvent.repository.owner.login,
-      url: githubEvent.repository.clone_url,
-      newHead: githubEvent.after,
-      urlSlug: urlSlug,
-      prefix: prefix,
-      project: project,
-      directory: directory,
-    },
+    payload: payload,
```
[the remark this answers is not in the document]
Comment (anabellabuckvar) [reply api/controllers/v2/github.ts]: Done

File: src/repositories/projectsRepository.ts (hunk 5 above)
Comment (anabellabuckvar) [reply src/repositories/projectsRepository.ts]: I added `projects` as a constant as done with the collections for the other repository classes except that I didn't configure it to pull the value from AWS parameter store. If that looks good to you both I'll do the same with `docs_metadata` as a constant

File: .github/workflows/deploy-stg-enhanced-webhooks.yml [hunk 10]
```diff
@@ -1,9 +1,10 @@
 on:
   push:
-    paths: ["api/v2/**", "cdk-infra/lib/constructs/api/**", "cdk-infra/utils/**"]
+    # paths: ["api/v2/**", "cdk-infra/lib/constructs/api/**", "cdk-infra/utils/**"]
```
Comment (mmeigs) [inline .github/workflows/deploy-stg-enhanced-webhooks.yml:3]: I believe put this back in?

File: .github/workflows/deploy-stg-enhanced-webhooks.yml [hunk 11]
```diff
@@ -1,9 +1,10 @@
...
     branches:
       - "main"
       - "integration"
+      - "DOP-4414"
```
Comment (mmeigs) [inline .github/workflows/deploy-stg-enhanced-webhooks.yml:7]: NOTE: reminder to remove before merge :)

File: api/controllers/v2/github.ts [hunk 12]
```diff
@@ -79,6 +133,139 @@ export const TriggerBuild = async (event: APIGatewayEvent): Promise<APIGatewayPr
...
+  //   return {
+  //     statusCode: 202,
+  //     headers: { 'Content-Type': 'text/plain' },
+  //     body:
+  //       `Build on branch
+  //       ${body.workflow_run.head_branch}
+  //       will not trigger site deployments as it was not on main branch in upstream repo`,
+  //   };
+  // }
+
```
Comment (mmeigs) [inline api/controllers/v2/github.ts:186]: Should this be removed or uncommented?

**Review by mmeigs — COMMENTED** [review]
I think the environment variable should work out, but you don't need to add it to any of the 4 config files you have because those are never touched if you're not accessing it through the config object!

File: .github/workflows/deploy-stg-enhanced-webhooks.yml (hunk 10 above)
Comment (anabellabuckvar) [reply .github/workflows/deploy-stg-enhanced-webhooks.yml:3]: Yup, this was just commented out for testing  purposes!

File: api/controllers/v2/github.ts (hunk 12 above)
Comment (anabellabuckvar) [reply api/controllers/v2/github.ts:186]: This will be uncommented before merge, it just needs to be commented out so I can continue testing any changes

File: api/controllers/v2/github.ts [hunk 13]
```diff
@@ -79,6 +133,133 @@ export const TriggerBuild = async (event: APIGatewayEvent): Promise<APIGatewayPr
...
+          return jobId;
+        } catch (err) {
+          consoleLogger.error('TriggerBuildError', `${err} Error inserting job for ${repoName}`);
+          return err;
+        }
+      })
+    );
+  }
+
+  let returnVal;
```
Comment (branberry) [inline api/controllers/v2/github.ts:228]: NIT: We don't need to use `let returnVal` here. We can return in the `try` statement with the `returnVal`

**Review by branberry — APPROVED** [review]
One small non-blocking change. Other than that, LGTM!

File: src/repositories/repoEntitlementsRepository.ts [hunk 14]
```diff
@@ -73,7 +73,7 @@ export class RepoEntitlementsRepository extends BaseRepository {
       `Mongo Timeout Error: Timedout while retrieving entitlements for ${slackUserId}`
     );
     // if user has specific entitlements
-    if ((entitlementsObject?.repos?.length ?? 0) > 0) {
+    if (entitlementsObject?.repos && entitlementsObject.repos.length > 0) {
```
Comment (mmeigs) [inline src/repositories/repoEntitlementsRepository.ts:76]: nit: we can shorten this to `entitlementsObject?.repos?.length > 0`

File: api/controllers/v2/github.ts [hunk 15]
```diff
@@ -1,71 +1,125 @@
...
+  let branchName: string;
+  let jobType: string;
+  let action: string;
+  let url: string;
+
+  if (isSmokeTestDeploy) {
+    url = 'https://github.com/' + repoOwner + '/' + repoName;
+    action = 'automatedTest';
+    jobType = 'productionDeploy';
+    branchName = 'master';
```
[the remark this answers is not in the document]
Comment (anabellabuckvar) [reply api/controllers/v2/github.ts:84]: Good point, thank you. I did consider this, but I checked at least the specific content repos that Allison requested always be deployed for smoke tests, and those at least all use "master" as the master branch.

If, at a later date, someone, for some reason, adds a repo to the list that uses main, they will have to add additional logic to handle that then. However, I agree that I don't believe it should be blocking right now for that hypothetical

File: src/job/productionJobHandler.ts [hunk 16]
```diff
@@ -141,24 +141,36 @@ export class ProductionJobHandler extends JobHandler {
...
+  //         return `${this.currJob.payload.prefix}/${this.currJob.payload.urlSlug}`;
+  //       }
+  //     }
+  //     return this.currJob.payload.prefix;
+  //   } catch (error) {
+  //     this.logger.save(this.currJob._id, error).then();
+  //     throw new InvalidJobError(error.message);
+  //   }
+  // }
+
```
[the remark this answers is not in the document]
Comment (anabellabuckvar) [reply src/job/productionJobHandler.ts]: It was commented out so I could do some testing! It's added back in now :)

File: api/controllers/v2/github.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (mmeigs) [reply api/controllers/v2/github.ts:84]: Agreed!

File: src/job/productionJobHandler.ts [hunk 17]
```diff
@@ -141,36 +141,24 @@ export class ProductionJobHandler extends JobHandler {
...
-      ? `${this.currJob.payload.prefix}/${this.currJob.payload.urlSlug}`
-      : this.currJob.payload.prefix;
-    if (this.currJob.payload.action == 'automatedTest') {
-      const titleArray = this.currJob.title.split(' ');
-      const commitHash = titleArray[titleArray.length - 5];
-      return `${prefix}/smokeTests/${commitHash}`;
+    try {
+      if (this.currJob.payload.prefix && this.currJob.payload.prefix === '') {
+        return this.currJob.payload.urlSlug ?? '';
+      }
```
Comment (mmeigs) [inline src/job/productionJobHandler.ts:147]: I'm only worried about this because it is in the productionJobHandler and it's a difference in pathing that does not solely rely on making sure this is a deploy_all situation. Could we also make sure that this has a conditional dependent on if this job is specifically a deploy all slack command?

Comment (anabellabuckvar) [reply src/job/productionJobHandler.ts:147]: When I click on the link to the version of the productionJobHandler file you reviewed, it shows that it is only reflecting changes from the past 4 commits- the changes on these lines within those commits are actually just reverting previous changes made. If you look at  those lines in src/job/productionJobHandler.ts [here](https://github.com/mongodb/docs-worker-pool/pull/1011/files#diff-ca78240af09f6070db1fcf5584b68368d38686db292057b80ab9469e204c42fc) you'll actually see that there aren't any differences in comparison to the main branch there!

Comment (mmeigs) [reply src/job/productionJobHandler.ts:147]: ohhhhhhh wonderful.
You're correct, I must have been looking at the "changes since I last looked"

**Review by mmeigs — APPROVED** [review]
